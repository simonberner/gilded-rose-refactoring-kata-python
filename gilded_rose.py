# -*- coding: utf-8 -*-

class GildedRose(object):

    def __init__(self, items):
        self.items = items

    def update_quality(self):
        for item in self.items:
            item.update_item()


class Item:
    def __init__(self, name, sell_in, quality):
        self.name = name
        self.sell_in = sell_in
        self.quality = quality

    def update_item(self):
        is_not_aged_brie = self.name != "Aged Brie"
        is_not_backstage_pass = self.name != "Backstage passes to a TAFKAL80ETC concert"
        is_not_sulfuras = self.name != "Sulfuras, Hand of Ragnaros"

        if is_not_aged_brie:
            if is_not_backstage_pass:
                if self.quality > 0:
                    if is_not_sulfuras:
                        self.quality = self.quality - 1
            else:
                if self.quality < 50:
                    self.quality = self.quality + 1
                    if not is_not_backstage_pass:
                        if self.sell_in < 11 and self.quality < 50:
                            self.quality = self.quality + 1
                        if self.sell_in < 6 and self.quality < 50:
                            self.quality = self.quality + 1
            # Updating sellIn when not Sulfuras
            if is_not_sulfuras:
                self.sell_in = self.sell_in - 1
            # Code smell: repeated conditional
            if self.sell_in < 0:
                # ifelse
                if is_not_backstage_pass:
                    if self.quality > 0:
                        if is_not_sulfuras:
                            self.quality = self.quality - 1
                else:
                    self.quality -= self.quality

        # it is aged_brie
        else:
            if self.quality < 50:
                self.quality = self.quality + 1
            self.sell_in = self.sell_in - 1
            if self.sell_in < 0 and self.quality < 50:
                self.quality = self.quality + 1

    def __repr__(self):
        return "%s, %s, %s" % (self.name, self.sell_in, self.quality)
