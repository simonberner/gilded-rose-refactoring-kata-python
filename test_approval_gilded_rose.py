from gilded_rose import Item, GildedRose
from approvaltests import verify_all_combinations, Options, DiffReporter
from approvaltests.reporters import PythonNativeReporter


# Check that sellIn and Quality degrades by 1 for a normal item
def test_update_quality():
    name = "normal item"
    sell_in = 10
    quality = 4

    verify_all_combinations(
        do_update_quality,
        [[str(name)], [int(sell_in)], [int(quality)]],
        options=Options().with_reporter(reporter=PythonNativeReporter())
    )


def do_update_quality(name, sell_in, quality):
    items = [Item(name, sell_in, quality)]
    gilded_rose = GildedRose(items)
    gilded_rose.update_quality()
    item = gilded_rose.items[0]
    return __print_item_as_string(item)


def __print_item_as_string(item):
    return item.name + ", sellIn: " + str(item.sell_in) + ", Quality: " + str(item.quality)
