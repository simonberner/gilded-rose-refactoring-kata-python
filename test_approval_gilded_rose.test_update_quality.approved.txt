args: ('normal item', 10, 4) => 'normal item, sellIn: 9, Quality: 3'
args: ('Aged Brie', 10, 4) => 'Aged Brie, sellIn: 9, Quality: 5'
